samza-core/src/main/java/org/apache/samza/storage/StorageManagerUtil.java

```diff
@@ -69,6 +72,12 @@ public class StorageManagerUtil {
   private static final String SST_FILE_SUFFIX = ".sst";
   private static final CheckpointV2Serde CHECKPOINT_V2_SERDE = new CheckpointV2Serde();
 
+  // Unlike checkpoint or offset files, side input offset file can have multiple writers / readers,
+  // since they are written during SideInputTask commit and copied to store checkpoint directory
+  // by TaskStorageCommitManager during regular TaskInstance commit (these commits are on separate run loops).
+  // We use a (process-wide) lock to ensure that such write and copy operations are thread-safe.
```

Comment:
That's right, needed because there's no other coordination/synchronization between the two run loops. It is used during commit whenever they read/write/copy side input offsets file.

To avoid deadlocks, lock is only held within a single method invocation, and the methods do not call each other. On second thought, I'll use a semaphore instead of a reentrant lock to document and enforce this.